### PR TITLE
Blog App: validations, model specs, and n+1 problems

### PR DESCRIPTION
**Pull request by Laylo309**
- Add the following validations:
  - For the User model:
    - Name must not be blank.
    - PostsCounter must be an integer greater than or equal to zero.
  - For the Post model:
    - Title must not be blank.
    - Title must not exceed 250 characters.
    - CommentsCounter must be an integer greater than or equal to zero.
    - LikesCounter must be an integer greater than or equal to zero.
- Add unit specs for all of the models' methods and validations.
- Add flash messages in the create actions in all controllers.
- N+1 problem is solved when fetching all posts and their comments for a user.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -2,6 +2,8 @@ class User < ApplicationRecord
   has_many :posts, foreign_key: :author_id
   has_many :comments, foreign_key: :author_id
   has_many :likes, foreign_key: :author_id
+  validates :name, presence: true
+  validates :posts_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
```
Comment (AshakaE) [inline app/models/user.rb]: Please note posts_counter must be an integer greater than zero only.

- [x] Kindly update this validation

File: app/models/post.rb [hunk 2]
```diff
@@ -1,15 +1,17 @@
...
-  # A method that updates the posts counter for a user.
-  def update_posts_counter
-    user.increment!(:posts_counter)
+  belongs_to :user, class_name: :User, foreign_key: 'author_id'
+  has_many :likes, dependent: :destroy
+  has_many :comments, dependent: :destroy
+  after_save :increment_counter
+  validates :title, presence: true
+  validates :text, presence: true, length: { maximum: 250 }
+  validates :comments_counter, :likes_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
```
Comment (AshakaE) [inline app/models/post.rb]: Please note comments_counter and likes_counter must be an integer greater than zero only.

- [x] Kindly update this validation

File: app/controllers/posts_controller.rb [hunk 3]
```diff
@@ -20,7 +20,11 @@ def create
 
     respond_to do |format|
       format.html do
-        redirect_to user_post_path(@post.user.id, @post.id) if @post.save
+        if @post.save
+          redirect_to user_post_path(@post.user.id, @post.id)
+        else
+          flash[:alert] = 'Error: post is not published'
+        end
```
Comment (AshakaE) [inline app/controllers/posts_controller.rb]: - [x] Kindly add flash message for successful posts

File: app/controllers/likes_controller.rb [hunk 4]
```diff
@@ -5,7 +5,11 @@ def create
 
     respond_to do |format|
       format.html do
-        redirect_to user_post_path(post.user.id, post.id) if like.save
+        if like.save
+          redirect_to user_post_path(post.user.id, post.id)
+        else
+          flash[:alert] = 'Error: like is not published'
+        end
```
Comment (AshakaE) [inline app/controllers/likes_controller.rb]: - [x] Kindly add flash message for successful likes

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -5,7 +5,11 @@ def create
 
     respond_to do |format|
       format.html do
-        redirect_to user_post_path(post.user.id, post.id) if comment.save
+        if comment.save
+          redirect_to user_post_path(post.user.id, post.id)
+        else
+          flash[:alert] = 'Error: comment is not published'
+        end
```
Comment (AshakaE) [inline app/controllers/comments_controller.rb]: - [x] Kindly add flash message for successful comments

**Review by AshakaE — CHANGES_REQUESTED** [review]
### Hi @Laylo309  🙋‍♀️, 

Good job so far! 👌

Your:
   - Git flow ✔️ 
   - Pull request 👍 
   - README 👌 
  
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


#  Required Changes ♻️ 🚧
_Check the comments under the review._

###  Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/comments_controller.rb (hunk 5 above)
Comment (Laylo309) [reply app/controllers/comments_controller.rb]: Done✔

File: app/controllers/likes_controller.rb (hunk 4 above)
Comment (Laylo309) [reply app/controllers/likes_controller.rb]: Done✔

File: app/controllers/posts_controller.rb (hunk 3 above)
Comment (Laylo309) [reply app/controllers/posts_controller.rb]: Done✔

File: app/models/post.rb (hunk 2 above)
Comment (Laylo309) [reply app/models/post.rb]: Done✔

File: app/models/user.rb (hunk 1 above)
Comment (Laylo309) [reply app/models/user.rb]: Done✔

**Review by AshakaE — APPROVED** [review]
### Hi @Laylo309   🙋‍♀️, 

# STATUS APPROVED ✔

Thanks for making the requested changes 👏 
   
Your project is complete!  There is nothing else to say other than... it's time to merge it 🔗 🚢 
Congratulations! 🎉 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers 🥂  and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**